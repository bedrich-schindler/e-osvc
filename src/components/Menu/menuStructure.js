import IconDashboard from '@material-ui/icons/Dashboard';
import routes from '../../routes';

export default [
  {
    icon: IconDashboard,
    path: routes.dashboard.path,
    title: routes.dashboard.title,
  },
];
