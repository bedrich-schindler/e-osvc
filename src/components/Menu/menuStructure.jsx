import Badge from '@material-ui/core/Badge';
import IconDashboard from '@material-ui/icons/Dashboard';
import IconClients from '@material-ui/icons/People';
import IconInvoices from '@material-ui/icons/Description';
import IconPayments from '@material-ui/icons/Payment';
import IconProjects from '@material-ui/icons/Business';
import IconTimeRecords from '@material-ui/icons/Timer';
import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import routes from '../../routes';

const StyledBadge = withStyles((theme) => ({
  badge: {
    backgroundColor: theme.palette.grey[600],
    border: `2px solid ${theme.palette.background.paper}`,
    padding: '0 4px',
    right: 2,
    top: 13,
  },
}))(Badge);

export default [
  {
    icon: IconDashboard,
    path: routes.dashboard.path,
    title: routes.dashboard.title,
  },
  { isDivider: true },
  {
    icon: IconClients,
    path: routes.clients.path,
    title: routes.clients.title,
  },
  {
    icon: IconProjects,
    path: routes.projects.path,
    title: routes.projects.title,
  },
  { isDivider: true },
  {
    icon: IconInvoices,
    path: routes.invoices.path,
    title: routes.invoices.title,
  },
  { isDivider: true },
  {
    icon: IconTimeRecords,
    path: routes.timeRecords.path,
    title: routes.timeRecords.title,
  },
  { isDivider: true },
  {
    icon: () => (
      <StyledBadge
        badgeContent="D"
        color="primary"
      >
        <IconPayments />
      </StyledBadge>
    ),
    path: routes.taxes.path,
    title: routes.taxes.title,
  },
  {
    icon: () => (
      <StyledBadge
        badgeContent="S"
        color="primary"
      >
        <IconPayments />
      </StyledBadge>
    ),
    path: routes.socialInsurance.path,
    title: routes.socialInsurance.title,
  },
  {
    icon: () => (
      <StyledBadge
        badgeContent="Z"
        color="primary"
      >
        <IconPayments />
      </StyledBadge>
    ),
    path: routes.healthInsurance.path,
    title: routes.healthInsurance.title,
  },
  {
    icon: () => (
      <StyledBadge
        badgeContent="N"
        color="primary"
      >
        <IconPayments />
      </StyledBadge>
    ),
    path: routes.sicknessInsurance.path,
    title: routes.sicknessInsurance.title,
  },
];
