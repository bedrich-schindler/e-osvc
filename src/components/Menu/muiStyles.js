export const SIDEBAR_WIDTH = 255;

export default (theme) => ({
  appBar: {
    transition: theme.transitions.create(['width', 'margin'], {
      duration: theme.transitions.duration.leavingScreen,
      easing: theme.transitions.easing.sharp,
    }),
    zIndex: theme.zIndex.drawer + 1,
  },
  appBarShift: {
    marginLeft: SIDEBAR_WIDTH,
    transition: theme.transitions.create(['width', 'margin'], {
      duration: theme.transitions.duration.enteringScreen,
      easing: theme.transitions.easing.sharp,
    }),
    width: `calc(100% - ${SIDEBAR_WIDTH}px)`,
  },
  drawer: {
    flexShrink: 0,
    left: 0,
    position: 'fixed',
    whiteSpace: 'nowrap',
    width: SIDEBAR_WIDTH,
    zIndex: 1000,
  },
  drawerClose: {
    transition: theme.transitions.create('width', {
      duration: theme.transitions.duration.leavingScreen,
      easing: theme.transitions.easing.sharp,
    }),
    [theme.breakpoints.down(768)]: {
      width: 0,
    },
    width: theme.spacing(7) + 1,
  },
  drawerOpen: {
    transition: theme.transitions.create('width', {
      duration: theme.transitions.duration.enteringScreen,
      easing: theme.transitions.easing.sharp,
    }),
    width: SIDEBAR_WIDTH,
  },
  toolbar: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'flex-end',
    padding: '0 8px',
    ...theme.mixins.toolbar,
  },
});
