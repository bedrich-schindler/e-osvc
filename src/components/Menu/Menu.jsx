import PropTypes from 'prop-types';
import classNames from 'classnames';
import React from 'react';
import Divider from '@material-ui/core/Divider';
import IconButton from '@material-ui/core/IconButton';
import IconChevronLeft from '@material-ui/icons/ChevronLeft';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import List from '@material-ui/core/List';
import { generate } from 'shortid';
import { withRouter } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';
import muiStyles from './muiStyles';
import menuStructure from './menuStructure';

export class Menu extends React.Component {
  constructor(props) {
    super(props);

    this.renderMenuItem = this.renderMenuItem.bind(this);
  }

  renderMenuItem(menuItem) {
    const { history } = this.props;

    if (menuItem.isDivider) {
      return <Divider key={generate()} />;
    }

    return (
      <ListItem
        button
        key={menuItem.path}
        onClick={() => history.push(menuItem.path)}
      >
        {menuItem.icon && (
          <ListItemIcon>
            <menuItem.icon />
          </ListItemIcon>
        )}
        <ListItemText primary={menuItem.title} />
      </ListItem>
    );
  }

  render() {
    const {
      classes,
      isMenuOpened,
      onCloseMenu,
    } = this.props;

    return (
      <Drawer
        className={classNames(classes.drawer, {
          [classes.drawerOpen]: isMenuOpened,
          [classes.drawerClose]: !isMenuOpened,
        })}
        classes={{
          paper: classNames({
            [classes.drawerOpen]: isMenuOpened,
            [classes.drawerClose]: !isMenuOpened,
          }),
        }}
        open={isMenuOpened}
        variant="permanent"
      >
        <div className={classes.toolbar}>
          <IconButton onClick={onCloseMenu}>
            <IconChevronLeft />
          </IconButton>
        </div>
        <Divider />
        <List>
          {menuStructure.map(this.renderMenuItem)}
        </List>
      </Drawer>
    );
  }
}

Menu.propTypes = {
  classes: PropTypes.object.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  isMenuOpened: PropTypes.bool.isRequired,
  onCloseMenu: PropTypes.func.isRequired,
};

export default withStyles(muiStyles, { withTheme: true })(withRouter(Menu));
