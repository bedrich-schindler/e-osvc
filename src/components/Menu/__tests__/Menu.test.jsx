
import React from 'react';
import sinon from 'sinon';
import {
  mount,
  shallow,
} from 'enzyme';
import {
  getShortIdStub,
  restoreShortIdStub,
} from '../../../../tests/testUtils';
import { Menu } from '../Menu';

describe('rendering', () => {
  let shortIdStub = null;

  beforeEach(() => {
    shortIdStub = getShortIdStub();
  });

  afterEach(() => {
    restoreShortIdStub(shortIdStub);
  });

  it('renders correctly', () => {
    const tree = shallow(
      <Menu
        classes={{}}
        history={{ push: () => {} }}
        isMenuOpened={false}
        onCloseMenu={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with opened menu', () => {
    const tree = shallow(
      <Menu
        classes={{}}
        history={{ push: () => {} }}
        isMenuOpened
        onCloseMenu={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('functionality', () => {
  let shortIdStub = null;

  beforeEach(() => {
    shortIdStub = getShortIdStub();
  });

  afterEach(() => {
    restoreShortIdStub(shortIdStub);
  });

  it('calls onCloseMenu when close menu button is clicked', () => {
    const spy = sinon.spy();
    const tree = mount(
      <Menu
        classes={{}}
        history={{ push: () => {} }}
        isMenuOpened
        onCloseMenu={spy}
      />,
    );

    tree.find('button').at(0).simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});
