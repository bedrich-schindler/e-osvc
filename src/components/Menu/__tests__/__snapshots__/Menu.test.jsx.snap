// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<WithStyles(ForwardRef(Drawer))
  className="undefined"
  classes={
    Object {
      "paper": "undefined",
    }
  }
  open={false}
  variant="permanent"
>
  <div>
    <WithStyles(ForwardRef(IconButton))
      onClick={[Function]}
    >
      <ChevronLeftIcon />
    </WithStyles(ForwardRef(IconButton))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <DashboardIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Přehled"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <PeopleIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Klienti"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <BusinessIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Projekty"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <DescriptionIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Faktury"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Daně"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Sociální pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Zdravotní pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Nemocenské pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Drawer))>
`;

exports[`rendering renders correctly with opened menu 1`] = `
<WithStyles(ForwardRef(Drawer))
  className=""
  classes={
    Object {
      "paper": "",
    }
  }
  open={true}
  variant="permanent"
>
  <div>
    <WithStyles(ForwardRef(IconButton))
      onClick={[Function]}
    >
      <ChevronLeftIcon />
    </WithStyles(ForwardRef(IconButton))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <DashboardIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Přehled"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <PeopleIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Klienti"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <BusinessIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Projekty"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <DescriptionIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Faktury"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(Divider)) />
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Daně"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Sociální pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Zdravotní pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <icon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Nemocenské pojištění"
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Drawer))>
`;
